### PR TITLE
Adding workflow to deploy databases to azure

### PR DESCRIPTION
**Pull request by JosephDSchwartz**
The workflow runs the flyway migrate command to run migrations against an Azure Database. I might have to update the set-vars job after https://github.com/im-practices/.github/pull/72 gets merged in.

### CODE REVIEW

File: workflow-templates/im-deploy-az-database.yml [hunk 1]
```diff
@@ -0,0 +1,176 @@
...
+jobs:
+  set-vars:
+    runs-on: ubuntu-latest # Force this to run on github-hosted runner by using a tag that does not exist on self-hosted runners
+
+    outputs:
+      ENVIRONMENT: ${{ steps.clean-env.outputs.mapped_input }} # To use this value: ${{ needs.set-vars.outputs.ENVIRONMENT }}
+
+    steps:
+      # To use this value: ${{ needs.set-vars.outputs.ENVIRONMENT }}
+      - name: Set ENVIRONMENT
```
Comment (hpractv) [inline workflow-templates/im-deploy-az-database.yml:40]: Should this be using the new set-environment-variables-by-scope action? 🤓

**Review by hpractv — CHANGES_REQUESTED** [review]
Abbreviated hold until we can get the `set-variables-by-scope` action in place. 😉

File: workflow-templates/im-deploy-az-database.yml [hunk 2]
```diff
@@ -0,0 +1,176 @@
+# Workflow Code: BetrayedCod_v1    DO NOT REMOVE
```
Comment (danielle-casella-adams) [inline workflow-templates/im-deploy-az-database.yml:1]: hehehehe

File: workflow-templates/im-deploy-az-database.yml [hunk 3]
```diff
@@ -0,0 +1,176 @@
...
+
+# Prerequisites
+# 1 - Make sure the az secrets have been added to the environment
+# 2 - Make sure the scm restrictions include the prod github runners (previously it had just octopus)
```
Comment (danielle-casella-adams) [inline workflow-templates/im-deploy-az-database.yml:5]: We should re-word for the db specific requirement
```suggestion
# 2 - Make sure the prod runners have access to the database with a 'azurerm_sql_virtual_network_rule' 
```

File: workflow-templates/im-deploy-az-database.yml [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+  # Each env has their own stakeholder approval environment.  If no required reviewers are set for
+  # that environment, the workflow will continue without requiring anyone to approve the deployment.
+  stakeholder-approval:
+    needs: [set-vars]
+    runs-on: ubuntu-latest # Force this to run on github-hosted runner by using a tag that does not exist on self-hosted runners
+    environment: '${{ needs.set-vars.outputs.ENVIRONMENT }} Stakeholder Approval'
+    steps:
+      - run: echo "The current environment is ${{ needs.set-vars.outputs.ENVIRONMENT }}"
+      - name: Approval Received
+        run: echo "Stakeholder approval was received"
```
Comment (danielle-casella-adams) [inline workflow-templates/im-deploy-az-database.yml:71]: Do we need a corresponding attestor approval block below this?  I'm not sure if teams require QAs to sign off on db deploys or not.
```suggestion
  # Each env has their own stakeholder approval environment.  If no required reviewers are set for
  # that environment, the workflow will continue without requiring anyone to approve the deployment.
  stakeholder-approval:
    needs: [set-vars]
    runs-on: ubuntu-latest # Force this to run on github-hosted runner by using a tag that does not exist on self-hosted runners
    environment: '${{ needs.set-vars.outputs.ENVIRONMENT }} Stakeholder Approval'
    steps:
      - run: echo "The current environment is ${{ needs.set-vars.outputs.ENVIRONMENT }}.  The branch/tag/sha is ${{ env.GITHUB_REF }}."
      - name: Approval Received
        run: echo "Stakeholder approval was received"
  
  # Each env has their own attestor approval environment (meant for QA Attestations).  If no required reviewers are set for
  # that environment, the workflow will continue without requiring anyone to approve the deployment.
  attestor-approval:
    needs: [set-vars]
    runs-on: ubuntu-latest # Force this to run on github-hosted runner by using a tag that does not exist on self-hosted runners
    environment: '${{ needs.set-vars.outputs.ENVIRONMENT }} Attestor Approval'
    steps:
      - run: |
          echo The current environment is ${{ needs.set-vars.outputs.ENVIRONMENT }}.  The branch/tag/sha is ${{ env.GITHUB_REF }}."
          echo "Attestor approval was received or no required reviewers were set for this environment."        
```

File: workflow-templates/im-deploy-az-database.yml [hunk 5]
```diff
@@ -0,0 +1,176 @@
...
+
+  # This job needs to run for all environments because deploy-az-db relies
+  # on it but the steps inside this job will only run for the Prod env.
+  validate-tag-for-prod-deploys:
+    needs: [set-vars, stakeholder-approval]
```
Comment (danielle-casella-adams) [inline workflow-templates/im-deploy-az-database.yml:76]: This potentially needs to be updated with attestor-approval too if you include it above
```suggestion
    needs: [set-vars, stakeholder-approval, attestor-approval]
```

File: workflow-templates/im-deploy-az-database.yml [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+      - run: echo "The current environment is ${{ steps.clean-env.outputs.mapped_input }}"
+
+  # Each env has their own stakeholder approval environment.  If no required reviewers are set for
+  # that environment, the workflow will continue without requiring anyone to approve the deployment.
+  stakeholder-approval:
+    needs: [set-vars]
+    runs-on: ubuntu-latest # Force this to run on github-hosted runner by using a tag that does not exist on self-hosted runners
+    environment: '${{ needs.set-vars.outputs.ENVIRONMENT }} Stakeholder Approval'
+    steps:
+      - run: echo "The current environment is ${{ needs.set-vars.outputs.ENVIRONMENT }}"
```
Comment (danielle-casella-adams) [inline workflow-templates/im-deploy-az-database.yml:69]:
```suggestion
      - run: echo "The current environment is ${{ needs.set-vars.outputs.ENVIRONMENT }}.  The branch/tag/sha is ${{ env.GITHUB_REF }}."
```

File: workflow-templates/im-deploy-az-database.yml [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+      # In this job, always checkout the default branch (not the branch that was provided as an input).  Also use
+      # fetch-depth: 0 to retrieve the history and tags so we can check if a tag is reachable from the default branch.
+      - name: Checkout Repository
+        if: needs.set-vars.outputs.ENVIRONMENT == 'prod'
+        uses: actions/checkout@v2
+        with:
+          ref: 'main' # TODO: verify the name of your default branch
+          fetch-depth: 0
+
+      - run: echo "The current environment is ${{ needs.set-vars.outputs.ENVIRONMENT }}"
```
Comment (danielle-casella-adams) [inline workflow-templates/im-deploy-az-database.yml:88]:
```suggestion
      - run: echo "The current environment is ${{ needs.set-vars.outputs.ENVIRONMENT }}.  The branch/tag/sha is ${{ env.GITHUB_REF }}."
```

File: workflow-templates/im-deploy-az-database.yml [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+        if: needs.set-vars.outputs.ENVIRONMENT == 'prod'
+        with:
+          tag: ${{ env.GITHUB_REF }}
+
+  deploy-az-db:
+    needs: [set-vars, validate-tag-for-prod-deploys]
+    runs-on: [self-hosted, ubuntu-20.04]
+    environment: ${{ needs.set-vars.outputs.ENVIRONMENT }}
+    steps:
+      - name: Checkout Repository
```
Comment (danielle-casella-adams) [inline workflow-templates/im-deploy-az-database.yml:100]: The first step in the job should be printing out the environment/tag


```suggestion
      - run: echo "The current environment is ${{ needs.set-vars.outputs.ENVIRONMENT }}.  The branch/tag/sha is ${{ env.GITHUB_REF }}."

      - name: Checkout Repository
```

File: workflow-templates/im-deploy-az-database.yml [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+          az logout
+          az cache purge
+          az account clear
+
+  update-deployment-board-and-send-teams-notification:
+    runs-on: ubuntu-latest # Force this to run on github-hosted runner by using a tag that does not exist on self-hosted runners
+    needs: [set-vars, deploy-az-db]
+    if: always()
+    steps:
+      - run: echo "The current environment is ${{ needs.set-vars.outputs.ENVIRONMENT }}.  The Tag is ${{ env.RELEASE_TAG }}."
```
Comment (danielle-casella-adams) [inline workflow-templates/im-deploy-az-database.yml:142]:
```suggestion
      - run: echo "The current environment is ${{ needs.set-vars.outputs.ENVIRONMENT }}.  The branch/tag/sha is ${{ env.GITHUB_REF }}."
```

File: workflow-templates/im-deploy-az-database.yml [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+
+      # https://github.com/im-practices/git-er-done/blob/main/actions/deployment-board.md
+      - name: Update Deployment Board
+        if: always()
+        uses: im-open/update-deployment-board@v1.0.2
+        with:
+          github-token: ${{ secrets.GITHUB_TOKEN }} # Special per-job token generated by GH for interacting with the repo
+          environment: ${{ needs.set-vars.outputs.ENVIRONMENT }}
+          board-number: '' # TODO: Add the automated deployment board number or remove if not using an automated deployment project board.
+          ref: ${{ env.RELEASE_TAG }}
```
Comment (danielle-casella-adams) [inline workflow-templates/im-deploy-az-database.yml:157]:
```suggestion
          ref: ${{ env.GITHUB_REF}}
```

File: workflow-templates/im-deploy-az-database.yml [hunk 11]
```diff
@@ -0,0 +1,176 @@
...
+    inputs:
+      branch-tag-sha:
+        description: The branch, tag or sha of the terraform that should be deployed.  For Production, only tags reachable by the default branch will be accepted.
+        required: true
+      environment:
+        # TODO: Update this description with the correct environments
+        description: 'The environment to deploy to - dev|qa|stage|prod|demo|uat'
+        required: true
+env:
+  GITHUB_REF: ${{ github.event.inputs.branch-tag-sha  }} # This is the branch/tag/sha that we'll be deploying
```
Comment (danielle-casella-adams) [inline workflow-templates/im-deploy-az-database.yml:23]: If you decide to change this to just tags, the name of this env should be updated to 'RELEASE_TAG' like it is in the other workflows and then all of the places that now reference the GITHUB_TAG would need to be updated.

File: workflow-templates/im-deploy-az-database.yml (hunk 4 above)
Comment (JosephDSchwartz) [reply workflow-templates/im-deploy-az-database.yml:71]: Probably doesn't hurt.

File: workflow-templates/im-deploy-az-database.yml [hunk 12]
```diff
@@ -0,0 +1,176 @@
...
+
+# TODO: Ensure each of the repo-level and env-level secrets used in this workflow have been populated by an admin in your repository.
+# TODO: Set up a deployment board if it does not already exist: https://github.com/im-practices/git-er-done/blob/main/actions/deployment-board.md
+
+name: Deploy Database to Azure
+on:
+  push:
+  workflow_dispatch:
+    inputs:
+      branch-tag-sha:
```
[the remark this answers is not in the document]
Comment (JosephDSchwartz) [reply workflow-templates/im-deploy-az-database.yml:15]: I'll add this to the list of updates I'm going to make once the variable substitution stuff gets merged in.

File: workflow-templates/im-deploy-az-database.yml [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+      - run: echo "The current environment is ${{ steps.clean-env.outputs.ENVIRONMENT }}.  The Tag is ${{ env.RELEASE_TAG }}."
+
+      - uses: im-open/is-tag-reachable-from-default-branch@v1.0.0
+        if: needs.set-vars.outputs.ENVIRONMENT == 'prod'
+        with:
+          tag: ${{ env.RELEASE_TAG }}
+
+  deploy-az-db:
+    needs: [set-vars, validate-tag-for-prod-deploys]
+    runs-on: [self-hosted, ubuntu-20.04]
```
Comment (danielle-casella-adams) [inline workflow-templates/im-deploy-az-database.yml:115]: I love that it's on linux!

Comment (JosephDSchwartz) [reply workflow-templates/im-deploy-az-database.yml:115]: Me too! I was really happy when I got that working 😃

**Pull request merged**